<a href="https://colab.research.google.com/github/TheMikeste1/cse380-notebooks/blob/master/ponder_and_prove_combinatorics_and_probability.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove Combinatorics and Probability
#### Due: Saturday, 6 February 2021, 11:59 pm.

## Collaborators:
Michael Hegerhorst - Author

In [ ]:
from math import factorial, gcd, isqrt, sqrt, e  # Math 🤪
from functools import reduce  # Folding
from typing import Any, Generator  # More explicit typing and descriptions
from time import time  # Timing my 24 hour code
from decimal import Decimal  # Better float precision

## Conjecture

A number-theoretic conjecture of combinatorial significance is the following:

$degree2({2n \choose n}) =$ the "bits-on count" (or population count, or Hamming weight) of $n$.

$degree2(m)$ is defined as the number (degree, exponent) of 2's in the prime factorization of $m$.

In other words, for any $m$, a positive integer, $m = 2^e \cdot o$ where $o$ is an odd positive integer (could be 1) and $e$ is a natural number, including zero --- which would be the case when $m$ is odd. It's the $e$ that is the $degree2$ of $m$.

~Another way to state this conjecture is that the number of 1's in the binary expansion of ${2n \choose n}$ for positive integer $n$ is equal to the number of 2's in the prime factorization of $n$.~

Another way to state this conjecture is that the number of 2's in the prime factorization of ${2n \choose n}$  for positive integer $n$ is equal to the number of 1's in the binary expansion of $n$.

Your task is to write Python code to test this conjecture for as many positive integers as you can. See the self-assessment for more details.

Note: a `bitsoncount` function can be a one-liner in Python: `return bin(x).count('1')`



### Response

#### Math for the Curious

##### Solving for the Degree Variables

$\forall m > 0 \; \exists e$ where $m = 2^e \cdot o$ where $o \equiv_2 1 \land o > 0$.

$$
m = 2^e \cdot o =>
\frac{m}{o} = 2^e => e = \log_2{\frac{m}{o}}
$$

$$
o = \frac{m}{2^e}
$$

##### Reducing ${2n}\choose n$

$$
\frac{(2n)!}{(2n - n)!n!} =
\frac{(2n)!}{n!^2}
$$

#### Code Library

In [ ]:
# This code was provided in class
def nCk(n, k):
  if k < 0 or k > n:
    return 0
  else:
    result = 1
    d = 1
    g = 1
    m = min(k, n - k)
    while (d <= m):
      g = gcd(result, d)
      result = n * (result // g)
      result = (result // (d // g))
      n -= 1
      d += 1
    return result

def nCk_wrapper_2nCn(n): return nCk(2 * n, n)

In [ ]:
def n2Cn(n: int) -> int : return factorial(2 * n) // (factorial(n) ** 2)  # See Reducing 2nCn

In [ ]:
n = 100000

In [ ]:
#%time nCk_wrapper_2nCn(n)
None  # This is here so it doesn't print the result

In [ ]:
#%time n2Cn(n)
None  # This is here so it doesn't print the result

My ```n2Cn``` function is consistently faster, so we'll use it instead of the provided code.

In [ ]:
def bits_on_count(x): return bin(x).count('1')

In [ ]:
def factor(n: int) -> Generator[int, Any, None]:
    while n % 2 == 0:  # Special case with 2
        yield 2
        n = n // 2
    if n == 1:
        return

    p = 3
    largest_p = isqrt(n)

    while p <= largest_p:
        while n % p == 0:
            n = n // p
            yield p
        if n == 1:  # If n == 1, we're done
            return
        p += 2  # This would be better replaced with a prime number generator
    yield n  # Yield any leftovers

#### Conjecture Test

##### Basic tests

In [ ]:
n = 2
print(list(factor(n2Cn(n))).count(2), '|', bits_on_count(n))

1 | 1


In [ ]:
n = 3
print(list(factor(n2Cn(n))).count(2), '|', bits_on_count(n))

2 | 2


In [ ]:
n = 4
print(list(factor(n2Cn(n))).count(2), '|', bits_on_count(n))

1 | 1


In [ ]:
failed = False
for n in range(0, 100):
  fac = list(factor(n2Cn(n))).count(2)
  bit = bits_on_count(n)
  if fac != bit:
    print(f"Failed on n = {n}")
    failed = True
    break
if not failed: print("Success!")

Success!


In [ ]:
# n = 10000
# %time list(factor(n2Cn(n))).count(2) == bits_on_count(n)
# n *= 2
# %time list(factor(n2Cn(n))).count(2) == bits_on_count(n)
# n *= 2
# %time list(factor(n2Cn(n))).count(2) == bits_on_count(n)
# n *= 2
# %time list(factor(n2Cn(n))).count(2) == bits_on_count(n)
# n *= 2
# %time list(factor(n2Cn(n))).count(2) == bits_on_count(n)
None

Wall time: 70.5 ms
Wall time: 252 ms
Wall time: 948 ms
Wall time: 3.67 s
Wall time: 14.3 s


This trend approximately follows $y = \frac{x^2}{1.75 \times 10^9}$, where $x = n$. This formula overestimates slightly.

In [ ]:
sqrt(86400 * 1.75 * 10 ** 9)

12296340.919151518

In 24 hours (86400 seconds), I should be able to verify $x = 12296341$ if I so desired.

##### 24-hour Test

In [ ]:
# n = 0
# start_time = time()
# while time() - start_time < 86400:  # 86400 is 24 hours in seconds
#   fac = list(factor(n2Cn(n))).count(2)
#   bit = bits_on_count(n)
#   if fac != bit:
#     print(f"Failed on n = {n}")
#   n += 1
# print(f"Verified {n} numbers in 24 hours")

I was able to verify 76220 numbers (from 0 to 76220) in 24 hours.

## Basic Probability Theory Question
A dark room contains two barrels. The first barrel is filled with green marbles, the second is filled with a half-and-half mixture of green and blue marbles. So there's a 100% chance of choosing a green marble from the first barrel, and a 50% chance of choosing either color in the second barrel. You reach into one of the barrels (it's dark so you don't know which one) and select a marble at random. It's green. You select another. It's green too. You select a third, a fourth, a fifth, etc. Green each time. What is the *minimum* number of marbles you need to select to *exceed* a probability of 99% that you are picking them out of the all-green barrel? (Note that there are enough marbles so that the answer does not depend on how many marbles are in the second barrel.)


### Response

The probability that we pick a green marble out of the 50-50 barrel is $\frac12 = 50\%$. When choosing two specifics colors in a row, we multiply the fractions, so two greens in a row is $\frac12 \cdot \frac12 = \frac14 = 25\%$. This results in the formula $y = \frac{1}{2^x}$.

Whereas each barrel is equally likely to be chosen, each barrel has a 50% chance of being the one we're reaching into, just like one barrel is 50-50 green-blue marbles. If we equate the probability of removing subsequent green marbles to the probability of the barrel being the 50-50 barrel, then in order to be 99% sure we're reaching into the pure barrel we simply need to solve $100\% - 99\% = 1\% = \frac{1}{2^x}$

$$
\frac{1}{100} = \frac{1}{2^x} => 2^x = 100 => x = \log_2{100} \approx 6.64 \approx 7
$$

In [ ]:
2 ** 7

128

In [ ]:
2 ** 6

64

## A Related But Deeper Basic Probability Theory Question
Take a deep breath. Suppose Shakespeare's account is accurate and Julius Caesar gasped "You too, Brutus" before breathing his last. What is the probability that you just inhaled a molecule that Julius Caesar exhaled in his dying breath?

Assume that after more than two thousand years the exhaled molecules are uniformly spread about the world and the vast majority are still free in the atmosphere. Assume further that there are $10^{44}$ molecules of air in the world, and that your inhaled quantity and Caesar's exhaled quantity were each about $2.2 \times 10^{22}$ molecules.
### Hint
If a number $x$ is small, then $(1 - x)$ is approximately equal to $e^{-x}$.


### Response

If Caesar's exhale was $2.2 \times 10^{22}$ molecules and there are $10^{44}$ molecules of air in the world, then each molecule has about an $x = 2.2 \times 10^{-22}\%$ chance of being one of those molecules (if they are evenly distributed). This means the chance of a molecule *not* being one of Caesar's is $(1 - x) \approx e^{-x}\%$.

In [ ]:
x = Decimal(2.2 * 10 ** 22) / (10 ** 44)
print(f"Chance of being Caesar's = {x:.23f} or {round(x, 23)}")

Chance of being Caesar's = 0.00000000000000000000022 or 2.2E-22


In [ ]:
not_x = Decimal(e) ** (-x)
print(f"Chance of _not_ being Caesar's \u2248 {not_x:.23f}")

Chance of _not_ being Caesar's ≈ 0.99999999999999999999978


From here, we simply exponentiate `not_x` by the number of choices we get ($2.2 \times 10^{22}$), and subtract that from 1 to get the chance that at least one *is* Caesar's:

$$
(e^{-2.2 \times 10^{-22}})^{2.2 \times 10^{22}} = e^{-4.84} \approx 0.0079
$$

In [ ]:
all_not_x = not_x ** Decimal(2.2 * 10 ** 22)
print(f"All not Caesar's: {all_not_x * 100:.2f}%")

All not Caesar's: 0.79%


In [ ]:
print(f"At least one Caesar's: {(1 - all_not_x) * 100:.2f}%")

At least one Caesar's: 99.21%


This means there is an amazing 99.21% chance that at least one molecule in the breath contains one of Caesar's last breath molecules.

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


### What is true of my experience in general?
(5 points each, 15 points total)
- [X] I had fun.
- [X] I learned something new.
- [X] I achieved something meaningful, or something I can build upon at a later time.

### What is true of my report on what I learned?
(5 points each, 25 points total)
- [X] I wrote a sufficient number of well-written sentences.
- [X] My report is free of "mechanical infelicities" (misspelled words, grammatical errors, punctuation errors, etc.).
- [X] I reported on any connections I found between this investigation and something I already know.
- [X] I reported who were and what contribution each of my collaborators made.
- [X] I reported on how many numbers I was able to verify with a time/computation budget of 24 hours (in a row).


### What is true about my answers?
(15 points each, 60 points total)
- [X] I figured out how to run a Python program continuously for at least 24 hours.
- [X] I refrained from printing out anything except the highest number I verified, knowing that printing just slows a program down.
- [X] I got the right answer for the first probability theory question.
- [X] I got the right answer for the second probability theory question.


## TA Comments

Excellent job!!! I thoroughly enjoyed reading through and grading your assignment.